packages/vhd-lib/parseVhdStream.js

```diff
@@ -13,124 +14,246 @@ const cappedBufferConcat = (buffers, maxSize) => {
+        let blockCount = 0
+        for (let blockCounter = 0; blockCounter < header.maxTableEntries; blockCounter++) {
+          const batEntrySector = buffer.readUInt32BE(blockCounter * 4)
+          // unallocated block, no need to export it
+          if (batEntrySector !== BLOCK_UNUSED) {
+            const batEntryBytes = batEntrySector * SECTOR_SIZE
+            // ensure the block is not before the bat
+            assert.ok(batEntryBytes >= batOffset + batSize)
+            this._index.push({
+              type: 'block',
+              id: blockCounter,
+              offset: batEntryBytes,
+              size: fullBlockSize,
+            })
+            blockCount++
+          }
         }
+        // sort again index to ensure block and parent locator are in the right order
+        this._index.sort((a, b) => a.offset - b.offset)
+        item.blockCount = blockCount
+        batFound = true
       }
-      // sort again index to ensure block and parent locator are in the right order
-      index.sort((a, b) => a.offset - b.offset)
-      item.blockCount = blockCount
-    } else if (type === 'block') {
-      item.bitmap = buffer.slice(0, bitmapSize)
-      item.data = buffer.slice(bitmapSize)
+      yield item
     }
+  }
+
+  async *blocks() {
+    while (this._index.length > 0) {
+      const item = this._index.shift()
+      const buffer = await this._read(item.offset, item.size)
 
-    yield item
+      item.bitmap = buffer.slice(0, this._bitmapSize)
+      item.data = buffer.slice(this._bitmapSize)
+      item.buffer = buffer
+      yield item
+    }
+    /**
+     * the second footer is at filesize - 512 , there can be empty spaces between last block
+     * and the start of the footer
+     *
+     * we read till the end of the stream, and use the last 512 bytes as the footer
+     */
+    const bufFooterEnd = await readLastSector(this._stream)
+    assert(this.#bufFooter.equals(bufFooterEnd), 'footer1 !== footer2')
   }
 
-  /**
-   * the second footer is at filesize - 512 , there can be empty spaces between last block
-   * and the start of the footer
-   *
-   * we read till the end of the stream, and use the last 512 bytes as the footer
-   */
-  const bufFooterEnd = await readLastSector(stream)
-  assert(bufFooter.equals(bufFooterEnd), 'footer1 !== footer2')
+  async *parse() {
+    yield* this.headers()
+    yield* this.blocks()
+  }
 }
 
-function readLastSector(stream) {
-  return new Promise((resolve, reject) => {
-    let bufFooterEnd = Buffer.alloc(0)
-    stream.on('data', chunk => {
-      if (chunk.length > 0) {
-        bufFooterEnd = cappedBufferConcat([bufFooterEnd, chunk], SECTOR_SIZE)
+// hybrid mode : read the headers from the vhd stream, and read the blocks from nbd
+class StreamNbdParser extends StreamParser {
+  #nbdClient = null
+  #concurrency = 16
+
+  constructor(stream, nbdClient = {}) {
+    super(stream)
+    this.#nbdClient = nbdClient
+  }
+
+  async _readBlockData(item) {
```

Comment:
Any reason for not using private fields (i.e. `#`)? :thinking:

Reply:
can't build with them 
```
[15:07:47] Plumber found unhandled error:
 SyntaxError: Unexpected character '#' (168:8) while parsing /home/florent/Documents/xen-orchestra/packages/vhd-lib/parseVhdStream.js while parsing file: /home/florent/Documents/xen-orchestra/packages/vhd-lib/parseVhdStream.js
 ```

Reply:
This syntax cannot be used at the moment due to `xo-web`.